stdlib/either.mli

```diff
@@ -0,0 +1,88 @@
+(*   Copyright 2019 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Either type.
+
+    Either is the simplest and most generic sum type. It is isomorphic
+    to the [result] type; it should be used when there is no implied
+    success/failure semantics in the choice of one of the constructors.
+
+    @since 4.11 *)
+
+(* Unlike [result], no [either] type is made available in Stdlib,
+   one needs to access [Either.t] explicitly:
+
+   - This type is less common in typical OCaml codebases,
+     which prefer domain-specific variant types whose constructors
+     carry more meaning.
+   - Adding this to Stdlib would raise warnings in existing codebases
+     that already use a constructor named Left or Right:
+     + when opening a module that exports such a name,
+       warning 45 is raised
+     + adding a second constructor of the same name in scope kicks
+       in the disambiguation mechanisms, and warning 41 may now
+       be raised by existing code.
+
+   If the use becomes more common in the future we can always
+   revisit this choice.
+*)
+
+type ('a, 'b) t = Left of 'a | Right of 'b (**)
+(** A value of [('a, 'b) Either.t] contains either a value of ['a]  or a value
+    of ['b] *)
+
+val left : 'a -> ('a, 'b) t
+(** [left v] is [Left v]. *)
+
+val right : 'b -> ('a, 'b) t
+(** [right v] is [Right v]. *)
+
+val get_left : ('a, 'b) t -> 'a option
+(** [get_left (Left v)] is [Some v], [get_left (Right _)] is [None] *)
+
+val get_right : ('a, 'b) t -> 'b option
+(** [get_right (Right v)] is [Some v], [get_right (Left _)] is [None] *)
+
+val swap : ('a, 'b) t -> ('b, 'a) t
+(** [swap (Left v)] is [Right v], and [swap (Right v)] is [Left v]. *)
+
+val map_left : ('a1 -> 'a2) -> ('a1, 'b) t -> ('a2, 'b) t
+(** [map_left f e] is [Left (f v)] if [e] is [Left v]
+    and [e] if [e] is [Right _]. *)
+
+val map_right : ('b1 -> 'b2) -> ('a, 'b1) t -> ('a, 'b2) t
+(** [map_right f e] is [Right (f v)] if [e] is [Right v] and [e] if [e] is [Left _]. *)
+
+val map_both : left:('a1 -> 'a2) -> right:('b1 -> 'b2) -> ('a1, 'b1) t -> ('a2, 'b2) t
+(** [map_both ~left ~right e] is [Left (left v)] if [e] is [Left v]
+    and [Right (right v)] if [e] is [Right v]. *)
+
+val fold : left:('a -> 'c) -> right:('b -> 'c) -> ('a, 'b) t -> 'c
+(** [fold ~left ~right (Left v)] is [left v], and
+    [fold ~left ~right (Right v)] is [right v]. *)
```

Comment:
Other modules usually have `iter` aswell (though personally I see it as a special case of `fold`).

Reply:
I started with it, but it was just an alias for `fold` so I wasn't sure what to do. Maybe I'll add it again. Could we have some extra opinions on this?

Reply:
I would like to have an `iter`. I think it conveys the intention better than a `fold` returning `unit`.

Reply:
Let's add `iter` back.

Reply:
I also added `for_all` for good measure.